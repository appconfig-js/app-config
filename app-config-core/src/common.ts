import { JsonObject, isObject } from '@app-config/utils';
import { logger } from '@app-config/logging';

export type KeyFormatter = (key: string, separator: string) => string;

/** Strategy used in 'app-config vars' for variable names */
export function camelToScreamingCase(key: string, separator: string = '_'): string {
  return key
    .replace(/([^A-Z]+)([A-Z][a-z])/g, `$1${separator}$2`)
    .replace(/([^0-9]+)([0-9]+)/g, `$1${separator}$2`)
    .replace(/-/g, separator)
    .toUpperCase();
}

/** Strategy used in 'app-config vars' to extract variable names from hierachy */
export function flattenObjectTree(
  obj: JsonObject,
  prefix: string = '',
  separator: string = '_',
  formatter: KeyFormatter = camelToScreamingCase,
): { [key: string]: string } {
  return Object.entries(obj).reduce((merged, [key, value]) => {
    const flatKey = `${prefix}${prefix ? separator : ''}${formatter(key, separator)}`;

    let flattenedObject;

    if (isObject(value)) {
      flattenedObject = flattenObjectTree(value, flatKey, separator, formatter);
    } else if (Array.isArray(value)) {
      const flattenedArray = value.reduce<JsonObject>((acc, val, ind) => {
        return Object.assign(acc, { [ind]: val });
      }, {});

      flattenedObject = flattenObjectTree(flattenedArray, flatKey, separator, formatter);
    } else {
      flattenedObject = {
        [flatKey]: value,
      };
    }

    return Object.assign(merged, flattenedObject);
  }, {});
}

/** Strategy for renaming keys, used for 'app-config vars' */
export function renameInFlattenedTree(
  flattened: { [key: string]: string },
  renames: string[] = [],
  keepOriginalKeys = false,
): typeof flattened {
  for (const rename of renames) {
    const matched = /^(.*)=(.*)$/.exec(rename);

    if (matched) {
      const [, renameFrom, renameTo] = matched;
      if (flattened[renameFrom]) {
        flattened[renameTo] = flattened[renameFrom]; // eslint-disable-line no-param-reassign

        if (!keepOriginalKeys) {
          delete flattened[renameFrom]; // eslint-disable-line no-param-reassign
        }
      } else {
        logger.warn(`A rename was used ('${rename}'), but no value was found.`);
      }
    } else {
      logger.warn(`A rename was used ('${rename}'), but not in correct format.`);
    }
  }

  return flattened;
}
