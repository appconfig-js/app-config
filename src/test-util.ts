import { join } from 'path';
import { dir } from 'tmp-promise';
import { outputFile, remove } from 'fs-extra';

export const withFakeFiles = async (
  files: [string, string][],
  cb: (dir: string) => Promise<void>,
) => {
  const { path: tmp } = await dir();
  const filenames = files.map(([name]) => join(tmp, name));

  await Promise.all(
    filenames.map(async (name, i) => {
      await outputFile(name, files[i][1]);
    }),
  );

  try {
    await cb(tmp);
  } finally {
    await remove(tmp);
  }
};
