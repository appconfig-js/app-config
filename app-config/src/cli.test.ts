import execa from 'execa';
import { join } from 'path';
import { isWindows } from './common';
import { withTempFiles } from './test-util';

const run = async (argv: string[], options?: execa.Options) =>
  execa('node', [join(__dirname, '..', './dist/cli.js'), ...argv], {
    stdout: 'pipe',
    stderr: 'ignore',
    stdin: 'ignore',
    ...options,
  });

describe('vars', () => {
  it('fails with no app-config', async () => {
    await expect(run(['vars', '-q'])).rejects.toThrow();
  });

  it('prints simple app-config file', async () => {
    const APP_CONFIG = JSON.stringify({ foo: true });
    const { stdout } = await run(['vars', '-q'], { env: { APP_CONFIG } });

    expect(stdout).toMatchSnapshot();
  });

  it('uses provided environment variable prefix', async () => {
    const APP_CONFIG = JSON.stringify({ foo: true });
    const { stdout } = await run(['vars', '-q', '--prefix', 'MY_CONFIG'], { env: { APP_CONFIG } });

    expect(stdout).toMatchSnapshot();
  });

  it('renames a variable', async () => {
    const APP_CONFIG = JSON.stringify({ foo: true });
    const { stdout } = await run(['vars', '-q', '--rename', 'APP_CONFIG_FOO=BAR'], {
      env: { APP_CONFIG },
    });

    expect(stdout).toMatchSnapshot();
  });

  it('aliases a variable', async () => {
    const APP_CONFIG = JSON.stringify({ foo: true });
    const { stdout } = await run(['vars', '-q', '--alias', 'APP_CONFIG_FOO=BAR'], {
      env: { APP_CONFIG },
    });

    expect(stdout).toMatchSnapshot();
  });

  it('filters list of variables with --only', async () => {
    const APP_CONFIG = JSON.stringify({ foo: true, bar: true });
    const { stdout } = await run(['vars', '-q', '--prefix=""', '--only', 'FOO'], {
      env: { APP_CONFIG },
    });

    expect(stdout).toMatchSnapshot();
  });

  it('prints only nonSecrets without --secrets option', async () => {
    await withTempFiles(
      {
        '.app-config.yml': `
          non-secret: true
        `,
        '.app-config.secrets.yml': `
          secret: true
        `,
      },
      async (inDir) => {
        const { stdout: withoutSecrets } = await run(['vars', '-q', '-C', inDir('.')]);
        expect(withoutSecrets).toMatchSnapshot();

        const { stdout: withSecrets } = await run(['vars', '-q', '--secrets', '-C', inDir('.')]);
        expect(withSecrets).toMatchSnapshot();
      },
    );
  });
});

describe('create', () => {
  it('fails with no app-config', async () => {
    await expect(run(['create'])).rejects.toThrow();
  });

  it('prints simple app-config file', async () => {
    const APP_CONFIG = JSON.stringify({ foo: true });
    const { stdout } = await run(['create', '-q'], { env: { APP_CONFIG } });

    expect(stdout).toMatchSnapshot();
  });

  it('prints YAML format', async () => {
    const APP_CONFIG = JSON.stringify({ foo: true });
    const { stdout } = await run(['create', '-q', '--format', 'yaml'], { env: { APP_CONFIG } });

    expect(stdout).toMatchSnapshot();
  });

  it('prints JSON format', async () => {
    const APP_CONFIG = JSON.stringify({ foo: true });
    const { stdout } = await run(['create', '-q', '--format', 'json'], { env: { APP_CONFIG } });

    expect(stdout).toMatchSnapshot();
  });

  it('prints JSON5 format', async () => {
    const APP_CONFIG = JSON.stringify({ foo: true });
    const { stdout } = await run(['create', '-q', '--format', 'json5'], { env: { APP_CONFIG } });

    expect(stdout).toMatchSnapshot();
  });

  it('prints TOML format', async () => {
    const APP_CONFIG = JSON.stringify({ foo: true });
    const { stdout } = await run(['create', '-q', '--format', 'toml'], { env: { APP_CONFIG } });

    expect(stdout).toMatchSnapshot();
  });

  it('prints raw format', async () => {
    const APP_CONFIG = JSON.stringify({
      anObject: {},
      aBoolean: true,
      aNumber: 83.2,
      aString: 'foo',
    });

    const { stdout: out1 } = await run(
      ['create', '-q', '--format', 'raw', '--select', '#/aBoolean'],
      { env: { APP_CONFIG } },
    );
    expect(out1).toMatchSnapshot();

    const { stdout: out2 } = await run(
      ['create', '-q', '--format', 'raw', '--select', '#/aNumber'],
      { env: { APP_CONFIG } },
    );
    expect(out2).toMatchSnapshot();

    const { stdout: out3 } = await run(
      ['create', '-q', '--format', 'raw', '--select', '#/aString'],
      { env: { APP_CONFIG } },
    );
    expect(out3).toMatchSnapshot();

    await expect(
      run(['create', '-q', '--format', 'raw', '--select', '#/anObject'], { env: { APP_CONFIG } }),
    ).rejects.toThrow();
  });

  it('can select a nested property', async () => {
    const APP_CONFIG = JSON.stringify({ a: { b: { c: true } } });
    const { stdout: nested1 } = await run(['create', '-q', '--format', 'json', '--select', '#/a'], {
      env: { APP_CONFIG },
    });

    expect(nested1).toMatchSnapshot();

    const { stdout: nested2 } = await run(
      ['create', '-q', '--format', 'json', '--select', '#/a/b'],
      {
        env: { APP_CONFIG },
      },
    );

    expect(nested2).toMatchSnapshot();

    const { stdout: nested3 } = await run(
      ['create', '-q', '--format', 'json', '--select', '#/a/b/c'],
      {
        env: { APP_CONFIG },
      },
    );

    expect(nested3).toMatchSnapshot();
  });

  it('fails with invalid property selector', async () => {
    const APP_CONFIG = JSON.stringify({ a: true });

    await expect(
      run(['create', '-q', '--format', 'json', '--select', '#/b'], {
        env: { APP_CONFIG },
      }),
    ).rejects.toThrow();
  });

  it('uses fileNameBase option', async () => {
    await withTempFiles(
      {
        'my-app.yml': `
          foo: bar
        `,
        'my-app.schema.yml': `
          type: object
          properties:
            foo:
              type: string
        `,
      },
      async (inDir) => {
        const { stdout } = await run(['create', '--fileNameBase=my-app', '-C', inDir('.')]);
        expect(stdout).toMatchSnapshot();

        const APP_CONFIG = JSON.stringify({ foo: true });
        await expect(
          run(['create', '--fileNameBase=my-app', '-C', inDir('.')], { env: { APP_CONFIG } }),
        ).rejects.toThrow();
      },
    );
  });

  it('uses environmentOverride option', async () => {
    await withTempFiles(
      {
        '.app-config.yml': `
          val:
            $env:
              default: 42
              prod: 88
        `,
      },
      async (inDir) => {
        const { stdout: defaultValue } = await run(['vars', '-q', '-C', inDir('.')]);
        expect(defaultValue).toMatchSnapshot();

        const { stdout: production } = await run([
          'vars',
          '-q',
          '-C',
          inDir('.'),
          '--environmentOverride=production',
        ]);

        expect(production).toMatchSnapshot();
      },
    );
  });

  it('uses environmentVariableName option', async () => {
    const MY_CONF = JSON.stringify({ foo: true });

    const { stdout } = await run(['create', '-q', '--environmentVariableName=MY_CONF'], {
      env: { MY_CONF },
    });

    expect(stdout).toMatchSnapshot();

    await expect(
      run(['create', '-q', '--environmentVariableName=MY_CONF'], { env: { APP_CONFIG: MY_CONF } }),
    ).rejects.toThrow();
  });
});

describe('create-schema', () => {
  it('fails with no app-config schema', async () => {
    await expect(run(['create-schema'])).rejects.toThrow();
  });

  it('prints simple app-config schema', async () => {
    await withTempFiles(
      {
        '.app-config.schema.yml': `
          type: object
          properties:
            foo: { type: boolean }
        `,
      },
      async (inDir) => {
        const { stdout } = await run(['create-schema', '-C', inDir('.')]);

        expect(stdout).toMatchSnapshot();
      },
    );
  });

  it('prints YAML format', async () => {
    await withTempFiles(
      {
        '.app-config.schema.yml': `
          type: object
          properties:
            foo: { type: boolean }
        `,
      },
      async (inDir) => {
        const { stdout } = await run(['create-schema', '-C', inDir('.'), '--format', 'yaml']);

        expect(stdout).toMatchSnapshot();
      },
    );
  });

  it('prints JSON format', async () => {
    await withTempFiles(
      {
        '.app-config.schema.yml': `
          type: object
          properties:
            foo: { type: boolean }
        `,
      },
      async (inDir) => {
        const { stdout } = await run(['create-schema', '-C', inDir('.'), '--format', 'json']);

        expect(stdout).toMatchSnapshot();
      },
    );
  });

  it('prints JSON5 format', async () => {
    await withTempFiles(
      {
        '.app-config.schema.yml': `
          type: object
          properties:
            foo: { type: boolean }
        `,
      },
      async (inDir) => {
        const { stdout } = await run(['create-schema', '-C', inDir('.'), '--format', 'json5']);

        expect(stdout).toMatchSnapshot();
      },
    );
  });

  it('prints TOML format', async () => {
    await withTempFiles(
      {
        '.app-config.schema.yml': `
          type: object
          properties:
            foo: { type: boolean }
        `,
      },
      async (inDir) => {
        const { stdout } = await run(['create-schema', '-C', inDir('.'), '--format', 'toml']);

        expect(stdout).toMatchSnapshot();
      },
    );
  });

  it('can select a nested property', async () => {
    await withTempFiles(
      {
        '.app-config.schema.yml': `
          definitions:
            Foo:
              type: object
              properties:
                foo: { type: boolean }
            Bar:
              type: object
              properties:
                bar: { type: boolean }
        `,
      },
      async (inDir) => {
        const { stdout: nested1 } = await run([
          'create-schema',
          '-C',
          inDir('.'),
          '--select',
          '#/definitions/Foo',
        ]);
        expect(nested1).toMatchSnapshot();

        const { stdout: nested2 } = await run([
          'create-schema',
          '-C',
          inDir('.'),
          '--select',
          '#/definitions/Bar',
        ]);
        expect(nested2).toMatchSnapshot();
      },
    );
  });

  it('fails with invalid property selector', async () => {
    await withTempFiles(
      {
        '.app-config.schema.yml': `
          type: object
        `,
      },
      async (inDir) => {
        await expect(
          run(['create-schema', '-C', inDir('.'), '--select', '#/prop']),
        ).rejects.toThrow();
      },
    );
  });
});

describe('nested commands', () => {
  it('fails with no app-config', async () => {
    await expect(run(['-q', '--', isWindows ? 'SET' : 'env'])).rejects.toThrow();
  });

  it('passes environment variables down', async () => {
    const APP_CONFIG = JSON.stringify({ foo: true });
    const { stdout } = await run(['-q', '--', isWindows ? 'SET' : 'env'], { env: { APP_CONFIG } });

    expect(stdout.includes('APP_CONFIG_FOO=true')).toBe(true);
  });

  it('uses prefix in environment variables', async () => {
    const APP_CONFIG = JSON.stringify({ foo: true });
    const { stdout } = await run(['-q', '-p', 'MY_CONFIG', '--', isWindows ? 'SET' : 'env'], {
      env: { APP_CONFIG },
    });

    expect(stdout.includes('MY_CONFIG_FOO=true')).toBe(true);
  });

  it('uses file format in main environment variable', async () => {
    const APP_CONFIG = JSON.stringify({ foo: true });
    const { stdout } = await run(['-q', '--format', 'json', '--', isWindows ? 'SET' : 'env'], {
      env: { APP_CONFIG },
    });

    expect(stdout.includes('APP_CONFIG={"foo":true}')).toBe(true);
  });

  it('passes APP_CONFIG_SCHEMA variable in', async () => {
    await withTempFiles(
      {
        '.app-config.yml': `
          foo: true
        `,
        '.app-config.schema.yml': `
          $schema: http://json-schema.org/draft-07/schema
          type: object
          properties:
            foo: { type: boolean }
        `,
      },
      async (inDir) => {
        const { stdout } = await run([
          '--format',
          'json',
          '-C',
          inDir('.'),
          '--',
          isWindows ? 'SET' : 'env',
        ]);

        expect(stdout.includes('APP_CONFIG={"foo":true}')).toBe(true);
        expect(
          stdout.includes(
            'APP_CONFIG_SCHEMA={"$schema":"http://json-schema.org/draft-07/schema","type":"object","properties":{"foo":{"type":"boolean"}}}',
          ),
        ).toBe(true);
      },
    );
  });
});
