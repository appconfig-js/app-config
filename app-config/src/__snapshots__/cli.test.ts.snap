// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`create can select a nested property 1`] = `"{\\"b\\":{\\"c\\":true}}"`;

exports[`create can select a nested property 2`] = `"{\\"c\\":true}"`;

exports[`create can select a nested property 3`] = `"true"`;

exports[`create prints JSON format 1`] = `"{\\"foo\\":true}"`;

exports[`create prints JSON5 format 1`] = `"{foo:true}"`;

exports[`create prints TOML format 1`] = `
"foo = true
"
`;

exports[`create prints YAML format 1`] = `
"foo: true
"
`;

exports[`create prints raw format 1`] = `"true"`;

exports[`create prints raw format 2`] = `"83.2"`;

exports[`create prints raw format 3`] = `"foo"`;

exports[`create prints simple app-config file 1`] = `
"foo: true
"
`;

exports[`create uses environmentOverride option 1`] = `"APP_CONFIG_VAL=42"`;

exports[`create uses environmentOverride option 2`] = `"APP_CONFIG_VAL=88"`;

exports[`create uses environmentVariableName option 1`] = `
"foo: true
"
`;

exports[`create uses fileNameBase option 1`] = `
"foo: bar
"
`;

exports[`create-schema can select a nested property 1`] = `
"type: object
properties:
  foo:
    type: boolean
"
`;

exports[`create-schema can select a nested property 2`] = `
"type: object
properties:
  bar:
    type: boolean
"
`;

exports[`create-schema prints JSON format 1`] = `"{\\"type\\":\\"object\\",\\"properties\\":{\\"foo\\":{\\"type\\":\\"boolean\\"}},\\"$schema\\":\\"http://json-schema.org/draft-07/schema#\\"}"`;

exports[`create-schema prints JSON5 format 1`] = `"{type:'object',properties:{foo:{type:'boolean'}},$schema:'http://json-schema.org/draft-07/schema#'}"`;

exports[`create-schema prints TOML format 1`] = `
"type = \\"object\\"
\\"$schema\\" = \\"http://json-schema.org/draft-07/schema#\\"

[properties.foo]
type = \\"boolean\\"
"
`;

exports[`create-schema prints YAML format 1`] = `
"type: object
properties:
  foo:
    type: boolean
$schema: 'http://json-schema.org/draft-07/schema#'
"
`;

exports[`create-schema prints simple app-config schema 1`] = `
"type: object
properties:
  foo:
    type: boolean
$schema: 'http://json-schema.org/draft-07/schema#'
"
`;

exports[`vars aliases a variable 1`] = `
"APP_CONFIG_FOO=true
BAR=true"
`;

exports[`vars filters list of variables with --only 1`] = `"FOO=true"`;

exports[`vars prints only nonSecrets without --secrets option 1`] = `"APP_CONFIG_NON_SECRET=true"`;

exports[`vars prints only nonSecrets without --secrets option 2`] = `
"APP_CONFIG_SECRET=true
APP_CONFIG_NON_SECRET=true"
`;

exports[`vars prints simple app-config file 1`] = `"APP_CONFIG_FOO=true"`;

exports[`vars renames a variable 1`] = `"BAR=true"`;

exports[`vars uses provided environment variable prefix 1`] = `"MY_CONFIG_FOO=true"`;
